designs/variables-for-userdata.md

```diff
@@ -0,0 +1,206 @@
+
+## User Stories
+
+* As a cluster manager using a `Custom` AMIFamily, I want to inject short-lived secrets (like an authentication token), which are embedded in Kubernetes resources, into my node's `userData` so that nodes can be securely bootstrapped, without drift concerns.
+* As a cluster manager, I still want Karpenter to detect and remediate drift when I make intentional changes to the main `userData` template, for situations such as node upgrades.
+
+## Background on `userData` and Drift
+
+Karpenter's drift feature ensures that nodes in a cluster conform to the desired state defined in their corresponding `NodeClaim` and `NodeClass`. For the AWS provider, the `EC2NodeClass` CRD defines AWS-specific configuration, including the `userData` that runs on EC2 instance initialization. The drift detection mechanism works by comparing the launch template associated with a running instance to the launch template that *would be* generated from the current `EC2NodeClass` spec.
+
+If a controller/user modifies the `spec.userData` field to update a hardcoded, time-sensitive credential, Karpenter will automatically detect and begin the drifing process.
+
+While this is correct for intentional configuration changes, it creates unneccessary churn for values that are rotated frequently, and for which cannot be pulled during node bootstrapping time.
+
+Today using supported AMI types such as `AL2`, you are able to provide passed-in userData in like shell scripts, which can retrieve credentials from Session Manager Parameter store. However, some AMI types disallow this arbitrary shell executions, which require hardcoding any credentials directly into the passed-in userData to the `EC2NodeClass` spec.
+
+## Not in Scope
+
+* This document is not proposing to be able to toggle on/off the static drift capability for user-defined fields in the `EC2NodeClass` spec. This would not solve the use case of rotating credentials in the `userData` field, since the field is an unstructured string and can have virtually any format.
+
+## Solutions
+
+### Option 1: Introduce Non-Drifting Template Variables
+
+This option introduces a new map field, `userDataVariables` to the `EC2NodeClass` spec for defining key-value pairs that can be substituted into the `userData` field. This field and its variables would be explicitly ignored by the hashing algorithms used in the drift detection mechanism. Variables are referenced in `userData` by bare token names (e.g., AUTH_TOKEN) using simple search-and-replace during launch template creation.
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: default
+spec:
+  amiFamily: Custom
+  # New field for non-drifting variables
+  userDataVariables:
+    AUTH_TOKEN: "rotated-secret-token-value"
```

Comment:
As mentioned in the WG, maybe there should be a more rich structure to signify which variables are driftable and which are not.